### PR TITLE
CCITCARBON-272 Added nested variable support for -vars-data

### PR DESCRIPTION
**Pull request by JunqiZhang0**
- Fix parsing problem when there is nested var use in variable files
- Added documentation and examples
- Added test

### CODE REVIEW

File: teflo/helpers.py [hunk 1]
```diff
@@ -1378,9 +1392,23 @@ def validate_render_scenario(scenario, config, temp_data_raw=()):
     temp_data = {}
     [temp_data.update(t) for t in temp_data_objs]
     temp_data.update(os.environ)
+
+    for item in temp_data.items():
+        if type(item[1]) is not str:
+            # import pdb; pdb.set_trace()
```
Comment (rujutashinde) [inline teflo/helpers.py:1398]: remove comment

File: teflo/helpers.py [hunk 2]
```diff
@@ -1378,9 +1392,23 @@ def validate_render_scenario(scenario, config, temp_data_raw=()):
...
+            for val in item[1]._items():
+                for key in val[0].keys():
+                    temp_data.update({item[0]: temp_data[key]})
+        elif item[1].__contains__('{{') and item[1].__contains__('}}'):
+            import pdb
```
Comment (rujutashinde) [inline teflo/helpers.py:1403]: place all imports in the beginning of the file

File: teflo/helpers.py [hunk 3]
```diff
@@ -1378,9 +1392,23 @@ def validate_render_scenario(scenario, config, temp_data_raw=()):
...
+            pdb.set_trace()
+            output = preprocyaml(item[1], temp_data)
+            temp_data.update({item[0]: output})
+
     try:
         data = yaml.safe_load(template_render(scenario, temp_data))
         # adding master scenario as the first scenario data stream
+        # import pdb;pdb.set_trace()
```
Comment (rujutashinde) [inline teflo/helpers.py:1411]: remove comment

File: teflo/helpers.py [hunk 4]
```diff
@@ -1404,6 +1432,7 @@ def validate_render_scenario(scenario, config, temp_data_raw=()):
     except yaml.YAMLError as e:
         # here raising yaml error to differentiate yaml issue is with main scenario
         raise e
+    # import pdb;pdb.set_trace()
```
Comment (rujutashinde) [inline teflo/helpers.py:1435]: remove comment

File: teflo/helpers.py [hunk 5]
```diff
@@ -1378,9 +1392,23 @@ def validate_render_scenario(scenario, config, temp_data_raw=()):
     temp_data = {}
     [temp_data.update(t) for t in temp_data_objs]
     temp_data.update(os.environ)
+
+    for item in temp_data.items():
```
Comment (rujutashinde) [inline teflo/helpers.py:1396]: Add some comments as to what is being done here

File: teflo/helpers.py [hunk 6]
```diff
@@ -1350,6 +1349,21 @@ def check_for_var_file(config):
     return var_file_list
 
 
+def preprocyaml(input, temp_data):
+
```
Comment (rujutashinde) [inline teflo/helpers.py:1353]: Add description for this method as to what it is doing

**Review by rujutashinde — COMMENTED** [review]
Couple of things:
1. Could you send me some test data that you tested this out with
2. Add this to one of the unit tests, as in change a unit test to work with variables which are recursively templated, e.g. maybe within test     <def test_valid_run_multiple_vars(mock_method, runner)> to use recursive templated data?
3. Edit the title of teh PR to add the CCITCARBON ticket number, so it is reflected in jira
4. If you are unable to add a unit test to test out this code, make sure you add this feature witin one of our integration job scenarios

File: teflo/helpers.py [hunk 7]
```diff
@@ -1350,6 +1349,104 @@ def check_for_var_file(config):
     return var_file_list
 
 
+def validate_brakets(input):
```
Comment (rujutashinde) [inline teflo/helpers.py:1352]: the spelling of brackets

File: docs/users/variable_data.rst [hunk 8]
```diff
@@ -102,3 +102,53 @@ The above example will look like
 .. code-block:: bash
 
     teflo run -s scenario.yml -t provision
+
+
+Nested Variable Usage
+----------------------
+Currently teflo supports nested variable using any of above methods
```
Comment (rujutashinde) [inline docs/users/variable_data.rst:109]: Currently teflo supports nested variable using any of **below** methods

change above to below

File: docs/users/variable_data.rst [hunk 9]
```diff
@@ -102,3 +102,53 @@ The above example will look like
...
+The nested variable can only be string after parsing
+For example:
+A nested variable can look like below:
+1. nested_var: "hello"
+2. nested_var: {{ hey }}
+3. nested_var: "hello{{ hey }}"
+
+
+
+You can
```
Comment (rujutashinde) [inline docs/users/variable_data.rst:121]: run make docs and check this output , as it is not seen in a proper format . It needs some spacing done correctly

**Review by rujutashinde — CHANGES_REQUESTED** [review]
I ran the tests manually which went through fine with lists and additional variable files in differnt locations.

Also squash the commits

File: docs/users/variable_data.rst [hunk 10]
```diff
@@ -102,3 +102,53 @@ The above example will look like
...
+1. Use multiple layer nested vars
```
Comment (rujutashinde) [inline docs/users/variable_data.rst:122]: run make docs and check this output , as it is not seen in a proper format . It needs some spacing done correctly

File: docs/users/variable_data.rst (hunk 8 above)
Comment (JunqiZhang0) [reply docs/users/variable_data.rst:109]: It should be above because I mean the documentation above

File: docs/users/variable_data.rst (hunk 9 above)
Comment (JunqiZhang0) [reply docs/users/variable_data.rst:121]: Thanks!

**Review by rujutashinde — APPROVED** [review]
LGTM

**Pull request merged**